src/ppx/ppx_eliom_utils.ml

```diff
@@ -0,0 +1,524 @@
+        | `Escaped_value _ | `Injection _ ->
+          Location.raise_errorf ~loc
+            "The syntax ~%% ... can not be nested."
+      end
+    | _ -> AM.default_mapper.expr mapper expr
+
+  let structure_item mapper str =
+    let loc = str.pstr_loc in
+    match str.pstr_desc with
+    | Pstr_extension (({txt=("server"|"shared"|"client")}, _), _) ->
+      Location.raise_errorf ~loc
+        "Sections are only allowed at toplevel."
+    | _ -> AM.default_mapper.structure_item mapper str
+
+  let signature_item mapper sig_ =
+    let loc = sig_.psig_loc in
+    match sig_.psig_desc with
+    | Psig_extension (({txt=("server"|"shared"|"client")}, _), _) ->
+      Location.raise_errorf ~loc
+        "Sections are only allowed at toplevel."
+    | _ -> AM.default_mapper.signature_item mapper sig_
+
+  let eliom_mapper context =
+    let context = ref (context :> Context.t) in
+    { Ast_mapper.default_mapper
+      with
+        Ast_mapper.
+
+        expr = eliom_expr context ;
+
+        (* Reject sections not at toplevel. *)
+        structure_item ;
+        signature_item ;
+    }
+
+
+  (** Toplevel translation *)
+  (** Switch the current context when encountering [%%server] (resp. shared, client)
+      annotations. Call the eliom mapper and [Pass.server_str] (resp ..) on each
+      structure item.
+  *)
+
+  let dispatch (server, shared, client) field context str =
+    let f = match context with
+      | `Server -> server | `Shared -> shared | `Client -> client
+    in
+    let m = eliom_mapper context in
+    f @@ (field m) m str
+
+  let dispatch_str c _mapper =
+    dispatch Pass.(server_str, shared_str, client_str)
+      (fun x -> x.AM.structure_item) c
+
+  let dispatch_sig c _mapper =
+    dispatch Pass.(server_sig, shared_sig, client_sig)
+      (fun x -> x.AM.signature_item) c
+
+  let toplevel_structure context mapper structs =
+    let f pstr =
+      let loc = pstr.pstr_loc in
+      match pstr.pstr_desc with
+      | Pstr_extension (({txt}, PStr strs), _)
+        when is_annotation txt ["shared.start"; "client.start" ;"server.start"] ->
+        if strs <> [] then
+          [ Str.extension ~loc @@ AM.extension_of_error @@ Location.errorf ~loc
+              "The %%%%%s extension doesn't accept arguments." txt ]
+        else ( context := Context.of_string txt ; [] )
+      | Pstr_extension (({txt}, PStr strs), _)
+        when is_annotation txt ["shared"; "client" ;"server"] ->
+        let c = Context.of_string txt in
+        flatmap (dispatch_str c mapper) strs
+      | _ ->
+        dispatch_str !context mapper pstr
+    in
+    let loc = {(Location.in_file !Location.input_name) with loc_ghost = true} in
+    open_eliom_pervasives :: Pass.prelude loc @ flatmap f structs @ Pass.postlude loc
+
+  let toplevel_signature context mapper sigs =
+    let f psig =
+      let loc = psig.psig_loc in
+      match psig.psig_desc with
+      | Psig_extension (({txt=("shared"|"client"|"server" as txt)}, PStr strs), _) ->
```

Comment:
@Drup, this should be `"shared.start"|"client.start"|"server.start"` (consistently with structures and the documentation), right?